python/cuvs/cuvs/neighbors/ivf_pq/ivf_pq.pyx

```diff
@@ -0,0 +1,430 @@
+                 add_data_on_build=True,
+                 conservative_memory_allocation=False):
+        self.params.n_lists = n_lists
+        self._metric = metric
+        self.params.metric = <DistanceType>DISTANCE_TYPES[metric]
+        self.params.metric_arg = metric_arg
+        self.params.kmeans_n_iters = kmeans_n_iters
+        self.params.kmeans_trainset_fraction = kmeans_trainset_fraction
+        self.params.pq_bits = pq_bits
+        self.params.pq_dim = pq_dim
+        if codebook_kind == "subspace":
+            self.params.codebook_kind = codebook_gen.PER_SUBSPACE
+        elif codebook_kind == "cluster":
+            self.params.codebook_kind = codebook_gen.PER_CLUSTER
+        else:
+            raise ValueError("Incorrect codebook kind %s" % codebook_kind)
+        self.params.force_random_rotation = force_random_rotation
+        self.params.add_data_on_build = add_data_on_build
+        self.params.conservative_memory_allocation = \
+            conservative_memory_allocation
+
+    @property
+    def metric(self):
+        return self._metric
+
+    @property
+    def metric_arg(self):
+        return self.params.metric_arg
+
+    @property
+    def add_data_on_build(self):
+        return self.params.add_data_on_build
+
+    @property
+    def n_lists(self):
+        return self.params.n_lists
+
+    @property
+    def kmeans_n_iters(self):
+        return self.params.kmeans_n_iters
+
+    @property
+    def kmeans_trainset_fraction(self):
+        return self.params.kmeans_trainset_fraction
+
+    @property
+    def pq_bits(self):
+        return self.params.pq_bits
+
+    @property
+    def pq_dim(self):
+        return self.params.pq_dim
+
+    @property
+    def codebook_kind(self):
+        return self.params.codebook_kind
+
+    @property
+    def force_random_rotation(self):
+        return self.params.force_random_rotation
+
+    @property
+    def add_data_on_build(self):
+        return self.params.add_data_on_build
+
+    @property
+    def conservative_memory_allocation(self):
+        return self.params.conservative_memory_allocation
+
+
+cdef class Index:
+    """
+    IvfPq index object. This object stores the trained IvfPq index state
+    which can be used to perform nearest neighbors searches.
+    """
+
+    cdef cuvsIvfPqIndex_t index
+    cdef bool trained
+
+    def __cinit__(self):
+        self.trained = False
+        check_cuvs(cuvsIvfPqIndexCreate(&self.index))
+
+    def __dealloc__(self):
+        check_cuvs(cuvsIvfPqIndexDestroy(self.index))
+
+    @property
+    def trained(self):
+        return self.trained
+
+    def __repr__(self):
+        return "Index(type=IvfPq)"
+
+
+@auto_sync_resources
+def build(IndexParams index_params, dataset, resources=None):
+    """
+    Build the IvfPq index from the dataset for efficient search.
+
+    Parameters
+    ----------
+    index_params : IndexParams object
+    dataset : CUDA array interface compliant matrix shape (n_samples, dim)
```

Comment:
Now that we're not just exposing a simple device-level primitive, I think it's time we start supporting host-based (e.g. numpy) arrays as well, even if that means automatically allocating device memory and copying from host to device internally in the python layer like we do in cuML. I think we can push that change to a future PR, but do you mind creating an issue for it at least?

Reply:
https://github.com/rapidsai/cuvs/issues/96